src/components/rotate-object-button.js

```diff
@@ -0,0 +1,334 @@
+      this.active = false;
+    } else if (this.mode === ROTATE_MODE.PUPPET) {
+      this.active = true;
+      this.puppet.exponent = data.exponent;
+      const { Ci, Ci_inverse, Oi } = this.puppet;
+      this.hand.getWorldQuaternion(Ci);
+      Ci_inverse.copy(Ci).inverse();
+      this.o.getWorldQuaternion(Oi);
+    }
+  },
+
+  tick() {
+    this.planarInfo.plane.material.visible =
+      this.active && (this.mode === ROTATE_MODE.FREE || this.mode === ROTATE_MODE.AXIS);
+
+    if (!this.active || this.el.systems.userinput.get(paths.actions.cursor.drop)) {
+      this.active = false;
+      return;
+    }
+
+    const { plane, normal, intersections, Pi, Pp, Pc, PiPc, PpPc, XYi } = this.planarInfo;
+
+    this.camera.matrixNeedsUpdate = true;
+    this.camera.updateMatrixWorld(true);
+    this.o.matrixNeedsUpdate = true;
+    this.o.updateMatrixWorld(true);
+    plane.matrixNeedsUpdate = true;
+    plane.updateMatrixWorld(true);
+
+    if (this.mode === ROTATE_MODE.PUPPET) {
+      // Find controller delta as a quaternion, then apply it to the object, snapping in fixed increments if desired:
+      //    Cc = Cd * Ci
+      // =>      Cd = Cc * Ci_inverse * Oi
+      // so that if we set Od = Cd then
+      //    Oc = Cd * Oi = Cc * Ci_inverse * Oi
+      // Snap to fixed angle increments by converting Cc to an Euler,
+      // restricting the angles using Math.floor, and converting back to a quaternion.
+      // TODO The quaternion's exponential varies with the "reactivity" of the object under the action of the controller. I know that for the natural numbers, the effect is to apply the "whole" action of the quaternion rather than a part, so I make a special case for exponent == 2.
+      // TODO: After sensitivity via exponential, would like to know how to adjust it on a per-axis basis. (But which axes?)
+      const { Cc, Cd, Ci_inverse, Oi, Oc, exponent } = this.puppet;
+      this.hand.updateMatrixWorld(true);
+      this.hand.getWorldQuaternion(Cc);
+      Cd.copy(Ci_inverse).premultiply(Cc);
+      Oc.copy(Oi);
+      Oc.premultiply(Cd);
+      if (exponent == 2) {
+        Oc.premultiply(Cd);
+      }
+      this.o.quaternion.copy(Oc);
+      this.o.matrixNeedsUpdate = true;
+      return;
+    }
+
+    this.o.getWorldPosition(plane.position);
+    this.camera.matrixWorld.decompose(this.v, plane.quaternion, this.v2);
+    const cameraToPlaneDistance = this.v.sub(plane.position).length();
+
+    intersections.length = 0;
+    const far = this.raycaster.far;
+    this.raycaster.far = 1000;
+    plane.raycast(this.raycaster, intersections);
+    this.raycaster.far = far;
+    const intersection = intersections[0]; // point
+    if (!intersection) return;
+
+    normal.set(0, 0, -1).applyQuaternion(plane.quaternion);
+
+    Pc.copy(intersection.point);
+    PiPc.copy(Pc).sub(Pi);
+    if (this.mode === ROTATE_MODE.FREE || this.mode === ROTATE_MODE.GARY) {
+      const boost = AFRAME.scenes[0].systems.userinput.get(paths.actions.boost);
```

Comment:
TODO : move this out of right here and fix the "FREE" / "GARY" confusion where the two modes are smushed messily together here and swapped with the "boost" modifier key.